Comm Layer Extensibility Improvements

fabian-hk opened the pull request:
This PR allows users of the communication layer to specify which tag should be used for the events. This makes the usage of the comm layer more flexible in higher layers because every layer can use its custom tag for the comm layer events and thereby also use different ``higher_layer_event_preds``.

fabian-hk committed:
Allow users to specify the tag of the events

qaphla reviewed (dismissed):
I have a few minor renaming comments, and believe that one typeclass instance is unneeded here, but it also shouldn't cause problems if it's left in.

Otherwise, this looks fine to me --- being able to have separate instances of the communication layer is likely to be a necessity going forward into more complex composed protocols, and this seems like a cleaner solution than turning the higher-layer predicates into a split function, as it keeps all of the split/merge functionality in one place, which is easier to manage.

fabian-hk commented on src/lib/communication/DY.Lib.Communication.Core.fst:
```diff
@@ -48,8 +45,12 @@ type communication_event =
 
-instance event_communication_event: event communication_event = {
-  tag = comm_layer_event_tag;
+class comm_layer_event_core_tag = {
```
I wouldn't say it is per protocol only one instance, but per layer. So, for example, the request-response functions use the comm layer core functions and define one instance for them, and a protocol that uses the comm layer core functions defines another instance.

fabian-hk commented on src/lib/communication/DY.Lib.Communication.RequestResponse.fst:
```diff
@@ -74,11 +74,23 @@ type communication_reqres_event =
+  tag: string;
+}
+
+instance default_comm_layer_event_reqres_tag: comm_layer_event_reqres_tag = {
```
My idea was that there is a default instance so that you don't have to worry about it if you don't need multiple instances, and if you want to use another instance, you just overwrite this at the protocol level, but I am not really sure how F* chooses the typeclass instance. So if you say that this doesn't make sense, I will remove the default instance.

qaphla replied:
I would tend to agree that this default instance should be removed, but also wonder how the instance selection works. Conceptually it is nice to only need to specify if multiple instances are used, but even if that does work currently, I wouldn't be confident in its stability.

fabian-hk committed:
Use tag also for crypto predicates

fabian-hk committed:
Remove state update predicate

fabian-hk committed:
Typed reqres data structures

fabian-hk committed:
Fix enable_reqres_comm_layer_lemmas

qaphla reviewed (changes requested):
My only remaining objection is to the default tag, specifically that it's unclear whether this will continue to work when multiple instances are needed (and also, even if it currently does, if it would continue to work in the future).

qaphla commented:
Interesting, it still showed up for me. Must have been viewing an older version of the PR.

qaphla committed:
Merge branch 'main' into fabian-hk/communication-layer-extensibility-improvement

TWal commented:
I am still puzzled by the use of typeclasses to store the tags. If in the dependencies of one file there are multiple instances of `comm_layer_core_tag` or `comm_layer_reqres_tag`, then it's not clear which one typeclass resolution will choose. Can we make it depend on some type? I see some implicit `#a` hanging around, this could be a good candidate?

qaphla commented:
> I am still puzzled by the use of typeclasses to store the tags. If in the dependencies of one file there are multiple instances of `comm_layer_core_tag` or `comm_layer_reqres_tag`, then it's not clear which one typeclass resolution will choose. Can we make it depend on some type? I see some implicit `#a` hanging around, this could be a good candidate?

You probably have a better sense than I do of how this works, but my impression is that the typeclass instance for a library should be hidden in that library, and so the only instance directly exposed in an analysis is the one for that analysis' own direct usage of the communication layer, meaning that there's a unique relevant one in scope. If that hiding doesn't take place, then I can see that being a problem. Maybe this only occurs with use of interfaces? I'm also uncertain what happens if A opens B and B opens C --- does A then also get access to everything in C, including using typeclasse instances in C for resolution?

It's not obvious to me what a natural dependency (type or otherwise) for this class is --- I believe that while there is a type parameter to many of the request/response functions, requests and responses won't necessarily be of the same type, and so a single parameter to determine what tag to use seems like it could lead to fragmentation, where a request and a response that are meant to be part of the same instance of the layer have different tags. Similarly for the core tag, but with the separate types of confidential-only, authenticated-only, and confidential+authenticated messages.

Doing this without a typeclass is of course possible, but then this seems to require explicitly passing the relevant tag in whenever calling one of the communication layer functions --- avoiding the need for this extra argument is the main motivation to use a typeclass here.

qaphla commented:
One possibility might be to parameterize the layer with a typeclass that contains both the tag and the collection of relevant types (request/response for that layer, the various message types for the core layer), as for instance in `map_types`.

fabian-hk committed:
Typed events; typeclass for layer configuration

fabian-hk committed:
Merge branch 'fabian-hk/communication-layer-extensibility-improvement' of github.com:REPROSEC/dolev-yao-star-extrinsic into fabian-hk/communication-layer-extensibility-improvement

fabian-hk committed:
Request-response layer tag in core functions; request-response layer initialization function; Cleanup

fabian-hk commented:
@TWal I have now created two typeclasses: one for the core of the communication layer and one for the request-response pair layer. Both typeclasses are parameterized with a type and contain a parser and a tag as suggested by @qaphla. This means that per instance of the communication layer, there can only be one type, but that limitation also existed before. I believe this should fix the issue you raised, right? There are a lot of changes in this PR now because the states and events are directly storing the payload instead of first parsing it to bytes. This improved the efficiency of some proofs. Furthermore, I did some renaming to clearly separate functions in the core of the communication layer and functions in the request-response layer.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Invariants.fst:
```diff
@@ -17,46 +17,52 @@ open DY.Lib.Communication.Core
+      sk_usage == long_term_key_type_to_usage (LongTermPkeKey (comm_layer_pkenc_tag a))  receiver /\
       (get_label tr msg) `can_flow tr` (comm_label sender receiver) /\
-      event_triggered tr sender (CommConfSendMsg sender receiver msg)
+      (match parse a msg with
+      | None -> False
+      | Some msg_parsed -> event_triggered tr sender (CommConfSendMsg sender receiver msg_parsed <: communication_core_event a))
+      
```
`DY.Lib.Comparse.DYUtils.parse_and_pred msg (fun msg_parsed -> event_triggered tr sender (CommConfSendMsg sender receiver msg_parsed)` does the same thing as this. It's not that much cleaner, though, since the predicate itself doesn't have a nice name.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Invariants.fst:
```diff
@@ -17,46 +17,52 @@ open DY.Lib.Communication.Core
+#pop-options
 
-val pke_crypto_predicates_communication_layer_and_tag:
+val pke_crypto_predicates_communication_layer_core_and_tag:
```
Nit: If the tag is also core-specific, I think `pke_crypto_predicates_and_tag_communication_layer_core` makes more sense here.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst:
```diff
@@ -52,33 +53,32 @@ val initialize_communication_proof:
-    is_knowable_by (join (principal_label sender) (principal_label receiver)) tr payload_bytes /\
-    event_triggered tr sender (CommConfSendMsg sender receiver payload_bytes)
+    is_public_key_for tr pk_receiver (LongTermPkeKey (comm_layer_pkenc_tag a)) receiver /\
+    is_well_formed a (is_knowable_by (join (principal_label sender) (principal_label receiver)) tr) payload /\
```
While we're at this, should this `join (principal_label sender) (principal_label receiver)` be replaced with `comm_label sender receiver`?

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst:
```diff
@@ -52,33 +53,32 @@ val initialize_communication_proof:
-    event_triggered tr sender (CommConfSendMsg sender receiver payload_bytes)
+    is_public_key_for tr pk_receiver (LongTermPkeKey (comm_layer_pkenc_tag a)) receiver /\
+    is_well_formed a (is_knowable_by (join (principal_label sender) (principal_label receiver)) tr) payload /\
+    event_triggered tr sender (CommConfSendMsg sender receiver payload <: communication_core_event a)
```
I wonder if it makes sense to wrap this up into a named predicate, both to avoid the need to write the type ascription everywhere and to ensure that the two `sender`s match everywhere (and that that doesn't go wrong when making changes in the future).

qaphla replied:
I'm also a bit surprised that the type ascription is needed to begin with, since I'd expect it to be inferred from the type of `payload`, but that's not really a problem, just unexpected.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst:
```diff
@@ -212,70 +210,71 @@ let receive_confidential_proof #invs #a tr higher_layer_preds comm_keys_ids rece
   (requires (
-    let payload_bytes = serialize a payload in
-    has_communication_layer_crypto_predicates /\
+    (Some? pk_receiver <==> Inr? payload) /\ 
+    (None? pk_receiver <==> Inl? payload) /\
```
Is there a reason to not work with `a`-typed payloads in both the encrypted and unencrypted cases?

qaphla replied:
Looks like this would be a more substantial change to the behaviour of `sign_message`. I guess in the end this is only internal to the library, so it doesn't matter that much.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst:
```diff
@@ -286,112 +285,106 @@ let send_authenticated_proof #invs #a tr higher_layer_preds comm_keys_ids sender
-      let Some msg_signed_t = parse com_message_t msg_bytes in
+    | Some payload -> (
+      let Some msg_signed_t = parse comm_message_t msg_bytes in
       let SigMessage msg_signed = msg_signed_t in
```
```suggestion
      let Some (SigMessage msg_signed) = parse comm_message_t msg_bytes in
```

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst:
```diff
@@ -286,112 +285,106 @@ let send_authenticated_proof #invs #a tr higher_layer_preds comm_keys_ids sender
-              | Some payload -> (
-                exists plain_payload nonce.
-                  payload.b == pke_enc pk_receiver nonce plain_payload /\
-                  event_triggered tr sender (CommConfAuthSendMsg sender receiver plain_payload)
-                )
+        match sign_input with
+        | Plain sender' receiver' payload_bytes' -> (
+          is_well_formed a (is_publishable tr) (Inl?.v payload) /\
+          (
+            event_triggered tr sender (CommAuthSendMsg sender (Inl?.v payload) <: communication_core_event a) \/ 
+            is_corrupt tr (long_term_key_label sender)
+          )
+        )
+        | Encrypted sender' receiver' payload_bytes' pk_receiver -> (
+          is_publishable tr (Inr?.v payload) /\
+          (
+            (
+              exists plain_payload nonce.
+                (Inr?.v payload) == pke_enc pk_receiver nonce plain_payload /\
+                (match parse a plain_payload with
+                | None -> False
+                | Some plain_payload_parsed -> event_triggered tr sender (CommConfAuthSendMsg sender receiver plain_payload_parsed <: communication_core_event a))
+            ) \/ (
+              is_corrupt tr (long_term_key_label sender)
```
Is it expected that we lose the sender/sender' connection here? Previously we had `cm.sender == sender` and similar for the receiver.)

It also seems like `is_corrupt tr (long_term_key_label sender)` could be commuted outside of the match, which saves a few lines, if nothing else.

fabian-hk replied:
I reinserted the sender/sender' and receiver/receiver' connection.

I cannot pull out the `is_corrupt tr (long_term_key_label sender)` because the `is_publishable` also has to hold if the sender is corrupt, and I am not sure how to cleanly pull that out with the `either` type used for the payload.

qaphla replied:
Yes, I see that now on looking closer --- I had misread the parenthesization at first, and so it looked like it was further on the outside than I expected.

qaphla commented on src/lib/communication/DY.Lib.Communication.Data.fst:
```diff
@@ -11,19 +11,21 @@ open DY.Lib.Comparse.Parsers
-  = mk_parseable_serializeable ps_com_send_byte
+instance parseable_serializeable_bytes_comm_send_byte: parseable_serializeable bytes comm_send_byte
+  = mk_parseable_serializeable ps_comm_send_byte
+  *)
```
The comment lines at the top and bottom of this seem out of place --- if this is no longer used, probably we should remove it entirely.

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Invariants.fst:
```diff
@@ -38,69 +44,97 @@ let aead_crypto_predicate_communication_layer #cusages = {
+
+(*** Core Crypto Predicates ***)
+
+val pke_crypto_predicates_communication_layer_reqres_and_tag:
```
As above, I'd move `and_tag` before the `communication_layer_reqres`.

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Invariants.fst:
```diff
@@ -38,69 +44,97 @@ let aead_crypto_predicate_communication_layer #cusages = {
+  pred_later = (fun tr1 tr2 prin sess_id st -> ());
+  pred_knowable = (fun tr prin sess_id st -> (
+    match st with
+    | ClientSendRequest {server; request; key} ->
+      assert(is_well_formed a (is_knowable_by (principal_typed_state_content_label prin (DY.Lib.State.Typed.tag #(communication_states a)) sess_id st) tr) request)
+    | ServerReceiveRequest {request; key} ->
+      assert(is_well_formed a (is_knowable_by (principal_typed_state_content_label prin (DY.Lib.State.Typed.tag #(communication_states a)) sess_id st) tr) request)
+    | ClientReceiveResponse {server; response; key} ->
+      assert(is_well_formed a (is_knowable_by (principal_typed_state_content_label prin (DY.Lib.State.Typed.tag #(communication_states a)) sess_id st) tr) response)
```
A fun thing that could be done here
```
| ClientSendRequest {server=_; request=payload; key}
| ServerReceiveRequest {request=payload; key}
| ClientReceiveResponse {server=_; response=payload; key} -> assert(is_well_formed a (is_knowable_by (principal_typed_state_content_label prin (DY.Lib.State.Typed.tag #(communication_states a)) sess_id st) tr) payload)
```
As long as the same variables are bound in each pattern, you can have several different patterns all go to the same handler. In this case, `payload` and `key` are bound in each pattern (though `key` isn't used).

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Invariants.fst:
```diff
@@ -140,48 +174,49 @@ let default_comm_reqres_higher_layer_event_preds (a:Type) {| parseable_serialize
     match com_msg_t with
     | RequestMessage {request; key} -> (
-      event_triggered tr client (CommClientSendRequest client server request key)
+      match parse a request with
+      | None -> False
+      | Some request_parsed -> event_triggered tr client (CommClientSendRequest client server request_parsed key <: communication_reqres_event a)
```
As before with `parse_and_pred` --- this comes up several more places too and should probably be handled uniformly.

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Invariants.fst:
```diff
@@ -192,21 +227,20 @@ let request_response_event_preconditions #cinvs = {
-let has_communication_layer_reqres_event_predicates #invs #a higher_layer_preds higher_layer_resreq_preds =
-  has_event_pred (event_predicate_communication_layer #invs.crypto_invs higher_layer_preds) /\
-  has_event_pred (event_predicate_communication_layer_reqres #invs.crypto_invs higher_layer_resreq_preds)
+let has_communication_layer_reqres_event_predicates #invs #a #config higher_layer_resreq_preds =
+  has_event_pred (event_predicate_communication_layer_core (comm_core_higher_layer_event_preds_reqres a)) /\
+  has_event_pred (event_predicate_communication_layer_reqres higher_layer_resreq_preds)
```
Note: It might make sense to have a general `has_communication_layer_reqres_predicates` that includes both the event and state predicates (and crypto, if applicable). Same thing for core if it doesn't already exist.

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Properties.fst:
```diff
@@ -21,44 +21,45 @@ open DY.Lib.Communication.RequestResponse.Invariants
-      (exists sid. state_was_set tr client sid (ClientSendRequest {server; request; key})) \/
-      (exists sid. state_was_set tr client sid (ClientReceiveResponse {server; response; key}))
+      (exists sid. state_was_set #(communication_states a) #(local_state_communication_layer_session a) tr client sid (ClientSendRequest {server; request; key} <: communication_states a)) \/
+      (exists sid. state_was_set #(communication_states a) #(local_state_communication_layer_session a) tr client sid (ClientReceiveResponse {server; response; key} <: communication_states a))
```
It seems like either the implicit arguments or the type ascription should be unnecessary here (and given that `request` and `response` are specified as `a`, I would expect neither to be needed).

No change needed, but another surprising thing.

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst:
```diff
@@ -243,28 +277,26 @@ let compute_response_message_proof #cinvs #a tr server key nonce request respons
-    | None -> False
-    | Some request -> higher_layer_preds.send_response tr server request response) /\
-    has_communication_layer_state_predicates /\
+    has_communication_layer_reqres_crypto_predicates a /\
+    has_communication_layer_reqres_event_predicates higher_layer_preds /\
+    higher_layer_preds.send_response tr server req_meta_data.request response /\
+    has_communication_layer_reqres_state_predicates a /\
```
Here, for example, having one predicate that includes all of crypto/event/state at once would be cleaner.

qaphla commented on src/lib/communication/DY.Lib.Communication.Data.fst:
```diff
@@ -89,15 +91,15 @@ instance parseable_serializeable_bytes_authenticated_data: parseable_serializeab
-  | SigMessage: signed_communication_message -> com_message_t
-  | RequestMessage: request_message -> com_message_t
-  | ResponseMessage: response_envelope -> com_message_t
+type comm_message_t =
+  | SigMessage: signed_communication_message -> comm_message_t
+  | RequestMessage: request_message -> comm_message_t
+  | ResponseMessage: response_envelope -> comm_message_t
```
Maybe for a future set of changes, but I wonder if it makes sense to also parameterize these messages (`comm_message_t` as well as `request_message`, `response_envelope`, `signed_communication_message`) with the type `a` of the payload. This might simplify some of the typeclass resolution elsewhere.

fabian-hk replied:
I think we should not do this because we need to add `(a:Type) {|comm_layer_core_config a|} (b:Type) {|comm_layer_reqres_config b|}`  to the `comm_message_t`, which creates a dependency on the `comm_layer_reqres_config` in the communication layer core.

qaphla replied:
> I think we should not do this because we need to add `(a:Type) {|comm_layer_core_config a|} (b:Type) {|comm_layer_reqres_config b|}` to the `comm_message_t`, which creates a dependency on the `comm_layer_reqres_config` in the communication layer core.

Yes, that makes sense as a reason not to. I think the core idea of having messages more parameterized might still have merit, but would take more thinking about how to avoid that dependency.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.fst:
```diff
@@ -60,110 +65,105 @@ instance event_communication_event: event communication_event = {
-  #a:Type -> {| parseable_serializeable bytes a |} ->
-  principal -> principal -> a -> option bytes -> bytes -> bytes -> bytes
+  #a:Type -> {|comm_layer_core_config a|} ->
+  principal -> principal -> payload:either a bytes -> pk_receiver:option bytes{(Some? pk_receiver <==> Inr? payload) /\ (None? pk_receiver <==> Inl? payload)} -> bytes -> bytes -> bytes
```
I wonder if it makes sense to bundle the payload and `pk_receiver` into a single type? Something of the form `either a (bytes & bytes)`, though maybe with a nicer name given to the right case and some form of record to extract things properly --- `either a ({payload:bytes; pk_receiver:bytes})` or so?

fabian-hk committed:
Feedback Klaas; Cleanup

fabian-hk committed:
Make get_response_label opaque; Add mk_comm_layer_response_nonce_labeled_proof; Fix communication layer trace printing

fabian-hk committed:
Merge branch 'main' of github.com:REPROSEC/dolev-yao-star-extrinsic into fabian-hk/communication-layer-extensibility-improvement

fabian-hk committed:
Fix missing #pop-options

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Invariants.fst:
```diff
@@ -18,42 +18,39 @@ open DY.Lib.Communication.Core
-      | Some msg_parsed -> event_triggered tr sender (CommConfSendMsg sender receiver msg_parsed <: communication_core_event a))
-      
+      parse_and_pred 
+        (fun msg_parsed -> event_triggered tr sender (CommConfSendMsg sender receiver msg_parsed <: communication_core_event a)) 
```
If this verifies as-is it's fine, but we should be aware that it might be beneficial to name this function to help the SMT solver --- e.g., something like

```
val comm_conf_send_event_triggered :
 #a:Type0 ->{|comm_layer_core_config a|} ->
 trace -> principal -> principal -> a -> Type0
let comm_conf_send_event_triggered tr sender receiver msg_parsed = event_triggered tr sender (CommConfSendMsg sender receiver msg_parsed <: communication_core_event a)
```
Of course, the disadvantage is then needing several different such functions for different message types, which is a bit bulky, so probably only useful if needed for the solver to tell that two instances of that function call are the same.

qaphla replied:
Oh, I see that this exists in `DY.Lib.Communication.Core.Lemmas` already from a change earlier. Maybe it makes sense to pull back to `.Invariants` and use in both places.

qaphla replied:
Resolved in latest version.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Invariants.fst:
```diff
@@ -18,42 +18,39 @@ open DY.Lib.Communication.Core
-      | None -> False
-      | Some payload -> event_triggered tr sender (CommAuthSendMsg sender payload <: communication_core_event a))
+      get_label tr (serialize a payload) `can_flow tr` public /\
+      event_triggered tr sender (CommAuthSendMsg sender payload <: communication_core_event a)
```
This is a nice bit of clean-up.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Invariants.fst:
```diff
@@ -67,24 +64,32 @@ let sign_crypto_predicates_communication_layer #cusages a #config = {
+      serialize_wf_lemma a (bytes_well_formed tr1) payload;
+      ()
+    )
+    | _ -> ()
```
It might be good to explicitly split out the two remaining branches as
```
| Some (Encrypted _ _ _ _) -> ()
| None -> assert(False)
```
to make explicit that we expect the `None` case to never be reached. This way, if it does become reachable in the future, we have a clear indicator that that's happening as soon as the change is made, rather than it only turning up later.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.fst:
```diff
@@ -133,24 +127,23 @@ let send_authenticated #a comm_keys_ids sender receiver payload =
   return (Some msg_id)
 
 #push-options "--ifuel 1 --fuel 0"
 //[@@ "opaque_to_smt"]
```
This reminds me that we should at some point fix `verify_message` not being opaque, but that's not for this PR.

qaphla replied:
Issue #119 created for this.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst:
```diff
@@ -77,11 +84,11 @@ val encrypt_message_proof:
   ))
   (ensures
-    is_publishable tr (encrypt_message pk_receiver nonce payload)
+    is_publishable tr (encrypt_message pk_receiver nonce payload) 
```
```suggestion
    is_publishable tr (encrypt_message pk_receiver nonce payload)
```

Just trimming out this seemingly accidentally added space.

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst:
```diff
@@ -208,54 +213,79 @@ let receive_confidential_proof #invs #a tr higher_layer_preds comm_keys_ids rece
-  assert(bytes_invariant tr signature);
-  serialize_wf_lemma comm_message_t (is_publishable tr) msg_signed;
-  ()
+  // The proof is duplicated here to improve the proof performance
+  match input with
+    | Inl payload -> (
+      let sig_input:signature_input a = Plain sender receiver payload in
+      let sig_input_bytes = serialize (signature_input a) sig_input in
+      serialize_wf_lemma (signature_input a) (is_publishable tr) sig_input;
+      let signature = sign sk_sender nonce sig_input_bytes in
+      assert(bytes_invariant tr signature); // Improves proof performance
+      let msg_signed = SigMessage {msg=sig_input_bytes; signature} in
+      serialize_wf_lemma comm_message_t (is_publishable tr) msg_signed;
+      ()
+    )
+    | Inr (payload, pk) -> (
+      let sig_input:signature_input a = Encrypted sender receiver payload pk in
+      let sig_input_bytes = serialize (signature_input a) sig_input in
+      serialize_wf_lemma (signature_input a) (is_publishable tr) sig_input;
+      let signature = sign sk_sender nonce sig_input_bytes in
+      assert(bytes_invariant tr signature); // Improves proof performance
+      let msg_signed = SigMessage {msg=sig_input_bytes; signature} in
+      serialize_wf_lemma comm_message_t (is_publishable tr) msg_signed;
+      ()
+    )
+#pop-options
```
Messy, but I don't see an immediately obvious solution to help performance. Probably some intermediate assertion can be proven in a split fashion and then let the rest be done the same way, but I'm not sure what that intermediate would be. Probably something like the sign_pred for sig_input_bytes (maybe plus `is_publishable tr`).

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst:
```diff
@@ -315,27 +344,30 @@ val verify_message_proof:
                 (match parse a plain_payload with
                 | None -> False
-                | Some plain_payload_parsed -> event_triggered tr sender (CommConfAuthSendMsg sender receiver plain_payload_parsed <: communication_core_event a))
+                | Some plain_payload_parsed -> comm_conf_auth_send_event_triggered tr sender receiver plain_payload_parsed)
```
Another place where `parse_and_pred` could be used (that I overlooked last time).

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst:
```diff
@@ -21,8 +21,72 @@ open DY.Lib.Communication.RequestResponse.Invariants
+  )
+let get_response_label_knowable_reverse #cinvs #a tr req_meta_data msg =
+  reveal_opaque (`%get_response_label) get_response_label;
+  ()
```
Does it make sense to merge these to
```
val get_response_label_knowable:
  {|cinvs:crypto_invariants|} ->
  #a:Type0 -> {|comm_layer_reqres_config a|} -> tr:trace ->
  req_meta_data:comm_meta_data a -> msg:a ->
  Lemma
  (requires
    cinvs.usages == default_crypto_usages
  )
  (ensures
    is_well_formed a (is_knowable_by (get_response_label tr req_meta_data) tr) msg <==>
    is_well_formed a (is_knowable_by (get_label tr req_meta_data.key) tr) msg
  )
let get_response_label_knowable #cinvs #a tr req_meta_data msg =
  reveal_opaque (`%get_response_label) get_response_label;
  ()
```

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst:
```diff
@@ -21,8 +21,72 @@ open DY.Lib.Communication.RequestResponse.Invariants
+  Lemma
+  (requires
+    (is_well_formed a (is_knowable_by (get_response_label tr req_meta_data) tr) request \/
+      (is_well_formed a (is_publishable tr) request /\ is_publishable tr req_meta_data.key))
```
It looks like this lemma should hold even with the weaker precondition
```
    (is_well_formed a (is_knowable_by (get_response_label tr req_meta_data) tr) request \/
      (is_well_formed a (is_publishable tr) request))
```
It also feels like it should be provable without revealing `get_response_label`, since in the first branch, the requires matches the ensures, and in the second branch, the request is publishable, and that should imply that it is knowable by any label. Might take a lemma call, though, in which case the reveal could be simpler.

qaphla replied:
This lemma appears, in fact, to not be needed anywhere else (it is called in one place, but that call can be removed without affecting the proof), so I have removed it.

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst:
```diff
@@ -21,8 +21,72 @@ open DY.Lib.Communication.RequestResponse.Invariants
+  (ensures
+    get_response_label tr1 req_meta_data == get_response_label tr2 req_meta_data
+  )
+  [SMTPat (get_response_label tr1 req_meta_data); SMTPat (tr1 <$ tr2)]
```
This might be over-aggressive without also guarding on `get_response_label tr2 req_meta_data`, but I'm not fully sure of how often `get_response_label tr1 req_meta_data` is expected to be called/in the proof context (or, for that matter, how often `get_response_label tr2 req_meta_data` is --- it could be that protecting it more makes it not trigger where it's needed).

qaphla replied:
Initial testing shows that in at least one place, this lemma is relied upon in a way that doesn't work with the stricter pattern.

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst:
```diff
@@ -143,24 +206,24 @@ let send_request_proof #invs #a tr comm_keys_ids higher_layer_preds client serve
   server:principal -> msg_id:timestamp ->
   Lemma
   (requires
+    invs.crypto_invs.usages == default_crypto_usages /\
```
I see why this is here (from the `get_response_label_knowable` preconditions, which then come from `get_response_label` using the default crypto usages), but it feels a bit suspect in that it essentially prevents us from using DH or KDFs while also using the communication layer.

qaphla replied:
#121 has been created for this, but it seems like a bigger problem than can be solved in this PR (and also already exists, and so is not being *created* by this PR). As such, I'm marking this as resolved.

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst:
```diff
@@ -183,43 +246,52 @@ let receive_request_proof #invs #a #config tr comm_keys_ids higher_layer_preds s
         // Triggers event_triggered_at_implies_pred
-        assert(event_triggered_at tr' i client (req_send_event client))
+        assert(event_predicate_communication_layer_reqres higher_layer_preds (prefix tr_recv i) client (req_send_event client));
+        assert(event_triggered_at tr_recv i client (req_send_event client))
```
Both of these being needed is surprising --- I would expect the second line to trigger `event_triggered_at_implies_pred` and give the first line immediately, and so if the first line asserts as-is, the second one seems not to add anything.

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst:
```diff
@@ -234,49 +306,80 @@ val mk_comm_layer_response_nonce_proof:
-      is_knowable_by (get_response_label tr_out req_meta_data) tr_out nonce
+      trace_invariant tr_out /\
+      is_knowable_by (get_response_label tr_out req_meta_data) tr_out nonce /\
+      get_label tr_out nonce `can_flow tr_out` get_label tr_out req_meta_data.key
```
```suggestion
      is_knowable_by (get_label tr_out req_meta_data.key) tr_out nonce
```
(though I wonder if it's needed at all to have both of these in the postcondition, since they can be interconverted with `get_response_label_knowable` if the issue with `cinvs` can be worked out --- that would also remove the need to reveal `get_response_label` in the proof, since it could then be replaced with a lemma call).

qaphla replied:
This edit also applies for `mk_comm_layer_response_nonce_labeled_proof`

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst:
```diff
@@ -234,49 +306,80 @@ val mk_comm_layer_response_nonce_proof:
+  Lemma
+  (requires
+    trace_invariant tr /\
+    has_communication_layer_reqres_state_predicate a /\
+    has_communication_layer_reqres_crypto_predicates a /\
```
Could be nice to merge (also in the unlabelled version).

qaphla commented on src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst:
```diff
@@ -234,49 +306,80 @@ val mk_comm_layer_response_nonce_proof:
+    has_communication_layer_reqres_crypto_predicates a /\
+    bytes_well_formed tr req_meta_data.key // Can be derived from CommServerReceiveRequest event
+  )
+  (ensures (
+    match mk_comm_layer_response_nonce_labeled req_meta_data usg prin tr with
+    | (None, tr_out) -> trace_invariant tr_out
+    | (Some nonce, tr_out) -> (
+      trace_invariant tr_out /\
+      is_knowable_by (get_response_label tr_out req_meta_data) tr_out nonce /\
+      get_label tr_out nonce `can_flow tr_out` get_label tr_out req_meta_data.key
+    )
```
Should the postcondition here also say something about the nonce being knowable by `prin`?

qaphla reviewed (commented):
I think there are a few things that should be at least discussed before merging, specifically the use of default crypto usages in `get_response_label`, but otherwise most changes appear minor.

qaphla committed:
Merge branch 'main' into fabian-hk/communication-layer-extensibility-improvement

qaphla commented on src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst:
```diff
@@ -40,8 +40,8 @@ let enable_core_comm_layer_lemmas preds =
-#push-options "--ifuel 2"
-val initialize_communication_core_proof:
+#push-options "--ifuel 2 --z3rlimit 25"
+val initialize_communication_proof:
```
```suggestion
val initialize_communication_core_proof
```

qaphla committed:
Update src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst

Commit:
Moves several communication helper predicates to Core.Invariants.

The predicates for send events being triggered now live in Invariants,
and can be used throughout the communication layer where needed, simplifying
presentation.

It might make sense to do the same for receive events, but these seem to come up
less in contexts where the payload needs to be parsed first.

Commit:
Fix up a missing pop-options

Commit:
Cleanup pass of comm core

Commit:
Cleanup of reqres lemmas

Commit:
Cleanup of a few more lemmas in reqres

qaphla commented:
This is now complete and builds locally for me (we'll have to see when GH actions are working again). I've moved #121 into an issue, since it's not related to this PR, and otherwise everything has been addressed. Of course, there are still places where proofs could be cleaned up (and maybe made faster, too, which would be nice), but that's a separate task.

fabian-hk committed:
Fix issue with FStar version d28626aee4fa1d3208b255846c47e7f860b20ae4

fabian-hk commented:
I also think this PR can be merged. @TWal do you want to look over it again?

fabian-hk committed:
Merge branch 'main' into fabian-hk/communication-layer-extensibility-improvement

fabian-hk committed:
Fix merge conflicts
